src/unused_finder/graph.rs

```diff
@@ -0,0 +1,104 @@
+use std::{
+    collections::{HashMap, HashSet},
+    rc::Rc,
+};
+
+use super::{
+    node_visitor::{ExportKind, ImportedItem},
+    unused_finder_visitor_runner::ImportExportInfo,
+};
+
+#[derive(Debug, Clone, Default)]
+pub struct GraphFile {
+    pub is_used: bool,
+    // pub package_name: String,
+    pub file_path: String,
+    pub unused_exports: HashSet<ExportKind>,
+    pub import_export_info: ImportExportInfo,
+}
+
+impl GraphFile {
+    pub fn mark_item_as_used(&mut self, item: &ImportedItem) -> bool {
+        let item = ExportKind::from(item);
+        self.unused_exports.remove(&item)
+    }
+}
+
+pub type Edge = (String, String, ImportedItem);
+
+#[derive(Default, Debug)]
+pub struct Graph {
+    pub files: HashMap<String, Rc<GraphFile>>,
+    pub edges: HashSet<Edge>,
+}
+
+impl<'a> Graph {
+    pub fn bfs(&mut self, entries: Vec<String>) -> Vec<String> {
+        let edges = self.get_edges(entries);
+        let mut new_used_stuff = false;
+        let new_entries = edges
+            .iter()
+            .filter_map(|(_, imported, item)| {
+                if let Some(file) = self.files.get_mut(imported) {
+                    if let Some(file) = Rc::get_mut(file) {
```

Comment:
Consider moving this to a locked Arc instead of an Arc and performing a par_iter here, see how it impacts performnace. My hunch is that it should be somewhat more performant since this is largely CPU work & the frontier set will quickly become large during a traversal

Reply:
Changed it to be an Arc, still need to experiment a little bit more on this, I think will work on this after the import resolver optimization